doc/programs/mp4live.1
.TH "mp4live" "1" "0.9" "Cisco Systems Inc." "MPEG4IP"
.SH "NAME"
.LP 
mp4live \- record and/or transmit video and/or audio
.SH "SYNTAX"
.LP 
mp4live [\fIoptions\fP]

.SH "DESCRIPTION"
.LP 
This program is a Linux audio/video capture utility that can capture and encode 
audio and video in real\-time. The results can be written to either an mp4
file, transmitted onto the network via either unicast or multicast, or both 
simultaneously! The audio is encoded with MP3, and the video with MPEG\-4 
Simple Profile which is OpenDivx compatible.
.LP 
The program has a Gtk based user interface which displays a preview of the encoded video and controls to adjust the settings of the functions of the program. Settings are remembered and stored in ~/.mp4live_rc.
.LP 
See the program README for detailed information on hardware and software requirements, known problems, and unknowns.
.SH "OPTIONS"
.LP 
.TP 
\fB\-file\fR <\fIconfig\-file\fP>
Use the configuration information store in <config\-file> instead of that in ~/.mp4live_rc
.TP 
\fB\-automatic\fR
Act as if the "Start" button were pressed immediately upon program startup. Whatever is appropriate to do based on the configuration at startup will be done.
.TP 
\fB\-headless\fR
Act as in automatic mode, and do not display any user interface.
.SH "FILES"
.LP 
\fI~/.mp4live_rc\fP 
.br 
\fI/dev/video\fP 
.br 
\fI/dev/dsp\fP
.SH "AUTHORS"
.LP 
David Mackie <dmackie@cisco.com>
.br 
Bill May <wmay@cisco.com>
.SH "SEE ALSO"
.LP 
gmp4player(1) mpeg4ip/server/mp4live/README


NEWS
12/17/01 Version 0.9.2
    Added VC 6.0 projects for windows
    Upgraded to SDL version 1.2.3
    Added shortcut keys for mp4player video window
    Enhanced player debug reporting
    Added RTP debugging messages in player format.
    Disable player sending RTCP when b=RR:0 is set in SDP
	Added man pages for mp4v2 library API
	Added hint track reading utlities to mp4v2 library
	Fixed bugs in SDP version of IOD
    Enhanced player decode performance for all but ISO codec.

11/7/01 Version 0.9.1
	Command line encoding tools now build and run on Windows
	Added a script, mp4v1regenerate, that can convert most old mp4 files 
		created with original mp4 library to corrected mp4 file format 
		generated by new mp4 library
	Fixed bad casts in new mp4 library when creating hint tracks
	Added man pages for primary programs: 
		gmp4player, mp4creator, mp4encode, mp4live 

11/6/01 Version 0.9
	New implementation of mp4 file library and utlities.
	New implementation of mp4 file creator/hinter.
    Added RTP over RTSP (TCP) mode to player.
    Added player "catch up" code for slower processors
	ISMA compliance fixes
	Fixed RFC 2250 compliance problem
	Updated to DSS 3.0.1
    Update to UCL RTP version 1.2.8

9/17/01 Version 0.8
    Port to FreeBSD, BSD/OS, Mac OS X (including Cocoa UI for Mac OS X)
    Conversion of ISO codec to use automake
    Update to SDL 1.2.2, including BSD/OS CD-ROM enhancements.
    Fix various interoperability bugs.
    Added SDP decode debugging.
    Added HTTP fetch capability of SDP files for player,
        including HTTP redirect.
    Fix player for non-44100Hz AAC streams.
    Added setting of debug level for messages 
        from HTTP, RTSP and SDP libraries from player UI.
    Changed player bytestream to throw ints instead of chars 
        when error occurs.
    Added frame fragmentation to ISMA audio RTP bytestream.
    Added search for MP3 files.
    Added config variables for receive RTP ports for player.
    Added better player support for systems with no audio.
    
8/1/01 Version 0.7
    Added live encoding program, mp4live
    Improved player synchronization of audio and video for OSS 
        supported platforms.
    Improved OpenDivX encoder performance
    Added interleaved audio RTP payload format support
    Fixed a number of bugs with mp4 file library
    Fixed compiling on windows platforms

5/7/01 Version 0.6.2
    Fixes so that package now builds on RedHat 6.1, 6.2, 7.0, and 7.1
    LAME MP3 encoder is now included
    Fixes to play Philips MP4 files correctly
    Enabled MMX routines for OpenDivX encoder (x86 architectures only)
    Extended Darwin PlaylistBroadcaster to control multicast IP TTL.
    Faac now handles UNIX paths correctly
    Fixed several bugs in mp4encode script

4/23/01 Version 0.6.1
    Fixed interoperability between OpenDivX and ISO MPEG-4 video codecs
    Fixed RTSP interoperability problem with Darwin Streaming Server 2.0
    Updated mp4encode script to simplify the encoding process
    Player now keeps history list
    Enabled some post-processing of video in the player
    Added mp4extract utility
    Fixed avi2raw to handle zero length video frames in AVI file

3/30/01 Version 0.6
    Inclusion of the OpenDivx MPEG-4 video encoder/decoder
    Inclusion of MP3 audio in addition to AAC audio
    Inclusion of the Apple Darwin Quicktime Server
    Support for the MP4 file format
    Support for multicast streaming
    Added fullscreen mode to player
    Windows port of the player

2/5/01 Version 0.5
    Initial public release


server/mp4live/mp4live.cpp
/*
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 * 
 * Software distributed under the License is distributed on an "AS
 * IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * rights and limitations under the License.
 * 
 * The Original Code is MPEG4IP.
 * 
 * The Initial Developer of the Original Code is Cisco Systems Inc.
 * Portions created by Cisco Systems Inc. are
 * Copyright (C) Cisco Systems Inc. 2000, 2001.  All Rights Reserved.
 * 
 * Contributor(s): 
 *		Dave Mackie		dmackie@cisco.com
 *		Bill May 		wmay@cisco.com
 */

#define DECLARE_CONFIG_VARIABLES 1
#include "mp4live.h"
#include "media_flow.h"
#include <getopt.h>

int main(int argc, char** argv)
{
	int rc;
	char* configFileName = NULL;
	bool automatic = false;
	bool headless = false;
	extern int nogui_main(CLiveConfig* pConfig);
	extern int gui_main(int argc, char**argv, CLiveConfig* pConfig);

	// command line parsing
	while (true) {
		int c = -1;
		int option_index = 0;
		static struct option long_options[] = {
			{ "automatic", 0, 0, 'a' },
			{ "file", 1, 0, 'f' },
			{ "headless", 0, 0, 'h' },
			{ "version", 0, 0, 'v' },
			{ NULL, 0, 0, 0 }
		};

		c = getopt_long_only(argc, argv, "af:hv",
			long_options, &option_index);

		if (c == -1)
			break;

		switch (c) {
		case 'a':
			automatic = true;
			break;
		case 'f':	// -f <config-file>
			configFileName = stralloc(optarg);
			break;
		case 'h':
			headless = true;
			break;
		case 'v':
			fprintf(stderr, "%s version %s\n", argv[0], VERSION);
			exit(0);
		case '?':
		default:
			fprintf(stderr, 
				"Usage: %s [-f config_file] [--automatic] [--headless]\n",
				argv[0]);
			exit(-1);
		}
	}

	char userFileName[MAXPATHLEN];
	char* home = getenv("HOME");
	if (home) {
		strcpy(userFileName, home);
		strcat(userFileName, "/.mp4live_rc");
	}

	CLiveConfig* pConfig = NULL;
	try {
		pConfig = new CLiveConfig(MyConfigVariables,
			sizeof(MyConfigVariables) / sizeof(SConfigVariable),
			home ? userFileName : NULL);

		if (configFileName) {
			pConfig->ReadFromFile(configFileName);
		} else {
			// read user config file if present
			pConfig->ReadDefaultFile();
		}

		pConfig->m_appAutomatic = automatic;
	} 
	catch (CConfigException* e) {
		delete e;
	}

	CVideoSource::InitialVideoProbe(pConfig);

	pConfig->Regenerate();

	// attempt to exploit any real time features of the OS
	// will probably only succeed if user has root privileges
	if (pConfig->GetBoolValue(CONFIG_APP_USE_REAL_TIME)) {
#ifdef _POSIX_PRIORITY_SCHEDULING
		// put us into the lowest real-time scheduling queue
		struct sched_param sp;
		sp.sched_priority = 1;
		if (sched_setscheduler(0, SCHED_RR, &sp) < 0) {
# ifdef DEBUG
			debug_message("Unable to set scheduling priority: %s",
				strerror(errno));
# endif
		}
#endif /* _POSIX_PRIORITY_SCHEDULING */
#ifdef _POSIX_MEMLOCK
		// recommendation is to reserve some stack pages
		u_int8_t huge[1024 * 1024];
		memset(huge, 1, sizeof(huge));

		// and then lock memory
		if (mlockall(MCL_CURRENT|MCL_FUTURE) < 0) {
# ifdef DEBUG
			debug_message("Unable to lock memory: %s", 
				strerror(errno));
# endif
		}
#endif /* _POSIX_MEMLOCK */
	}

#ifdef NOGUI
	rc = nogui_main(pConfig);
#else
	if (headless) {
		rc = nogui_main(pConfig);
	} else {
		// hand over control to GUI
		try {
			rc = gui_main(argc, argv, pConfig);
		}
		catch (void* e) {
			rc = -1;
		}
	}
#endif

	// save any changes to user config file
	if (configFileName == NULL) {
		pConfig->WriteDefaultFile();
	} else {
		free(configFileName);
	}

	if (pConfig->GetBoolValue(CONFIG_APP_USE_REAL_TIME)) {
#ifdef _POSIX_MEMLOCK
		munlockall();
#endif /* _POSIX_MEMLOCK */
	}

	delete pConfig;
	exit (rc);
}

int nogui_main(CLiveConfig* pConfig)
{
	if (!pConfig->GetBoolValue(CONFIG_AUDIO_ENABLE)
	  && !pConfig->GetBoolValue(CONFIG_VIDEO_ENABLE)) {
		return -1;
	}

	// override any configured value of preview
	bool previewValue =
		pConfig->GetBoolValue(CONFIG_VIDEO_PREVIEW);
	pConfig->SetBoolValue(CONFIG_VIDEO_PREVIEW, false);

	CAVMediaFlow* pFlow = new CAVMediaFlow(pConfig);

	pFlow->Start();

	sleep(pConfig->GetIntegerValue(CONFIG_APP_DURATION)
		* pConfig->GetIntegerValue(CONFIG_APP_DURATION_UNITS));

	pFlow->Stop();

	delete pFlow;

	pConfig->SetBoolValue(CONFIG_VIDEO_PREVIEW, previewValue);

	return 0;
}
